app/screens/TransactionAuthorization.tsx

```diff
@@ -120,49 +142,63 @@ export function TransactionAuthorization (): JSX.Element | null {
         wallet !== undefined // just in case any data stuck in store
       ) {
         let result: CTransactionSegWit | null | undefined
-
         signTransaction(transaction, wallet.get(0), onRetry)
-          .then(async signedTx => { result = signedTx }) // positive
+          .then(async signedTx => {
+            result = signedTx
+          }) // positive
           .catch(e => {
             // error type check
-            if (e.message === 'invalid hash') result = null // negative, invalid passcode
-            // else result = undefined // neutral
+            if (e.message === INVALID_HASH) {
+              result = null // negative, invalid passcode
+            } else if (e.message !== USER_CANCELED) {
+              dispatch(ocean.actions.setError(e))
+            }
           })
           .then(async () => {
             // result handling, 3 cases
             if (result === undefined) {
               dispatch(transactionQueue.actions.pop())
             } else if (result === null) {
+              setAttemptsRemaining(MAX_PASSCODE_ATTEMPT)
+              await PasscodeAttemptCounter.set(0)
               await clearWallets()
               onUnlinkWallet()
             } else {
               await onComplete(dispatch, result)
             }
           })
           .catch(e => Logging.error(e))
-          .finally(() => emitEvent('IDLE'))
+          .finally(() => {
+            onSubmitCompletion()
+          })
       } else if (authentication !== undefined) {
-        let invalidPassphrase = false
+        let status = ''
         authenticateFor(onPrompt, authentication, onRetry)
           .catch(e => {
             // error type check
-            if (e.message === 'invalid hash') invalidPassphrase = true
-            // else result = undefined // neutral
+            if (e.message === INVALID_HASH) {
+              status = INVALID_HASH
+            } else if (e.message === USER_CANCELED) {
+              status = USER_CANCELED
+            } else if (e.message !== USER_CANCELED) {
+              dispatch(ocean.actions.setError(e))
+            }
           })
           .then(async () => {
-            if (invalidPassphrase) {
+            if (status === INVALID_HASH) {
+              setAttemptsRemaining(MAX_PASSCODE_ATTEMPT)
+              await PasscodeAttemptCounter.set(0)
               await clearWallets()
               onUnlinkWallet()
-            } else {
+            } else if (status !== USER_CANCELED) {
```

Comment:
please use the fix in #458 (which is more intuitive, do not have to wait for error preprocess for valid passcode callback)

Reply:
ah yeah make sense, let me try it locally first

Reply:
Updated @ivan-zynesis